libsolidity/ast/Types.cpp

```diff
@@ -199,18 +148,78 @@ util::Result<TypePointers> transformParametersToExternal(TypePointers const& _pa
 	return transformed;
 }
 
+bool compareTypes(Type const* lhs, Type const* rhs)
+{
+	solAssert(lhs && rhs, "");
+	return lhs->richIdentifier() < rhs->richIdentifier();
 }
+using TypeSet = std::set<Type const*, decltype(compareTypes)*> ;
 
-vector<frontend::Type const*> solidity::frontend::oversizedSubtypes(frontend::Type const& _type)
+void storageOversizedSubtypesInner(
+	Type const* _type,
+	bool _includeType,
+	set<string>& _seen,
+	TypeSet& _oversizedSubtypes
+)
 {
-	set<StructDefinition const*> structsSeen;
-	TypeSet oversized;
-	oversizedSubtypesInner(_type, false, structsSeen, oversized);
-	vector<frontend::Type const*> res;
+	for (Type const* subtype: _type->subtypes())
+	{
+		string id = subtype->richIdentifier();
+		if (_seen.count(id))
+			continue;
+		_seen.insert(id);
+		bool includeType = _type->category() != Type::Category::Struct;
+		storageOversizedSubtypesInner(subtype, includeType, _seen, _oversizedSubtypes);
+		_seen.erase(id);
+	}
+
+	if (_includeType && _type->storageSizeUpperBound() >= bigint(1) << 64)
+		_oversizedSubtypes.insert(_type);
+}
+
+}
+
+vector<Type const*> Type::storageOversizedSubtypes() const
+{
+	set<string> seen;
+	TypeSet oversized(compareTypes);
+	storageOversizedSubtypesInner(this, false, seen, oversized);
+	vector<Type const*> res;
 	copy(oversized.cbegin(), oversized.cend(), back_inserter(res));
 	return res;
```

Comment:
`TypeSet` with its home-brewed comparator is rather an implementation detail. I did not want to pollute the public interface with a yet another custom type.

Reply:
How about returning a plain set then? I still makes more sense in my opinion because the order does not matter and I don' think duplicates are desirable.

Reply:
`set` makes more sense, semantically. My small concern here is that `set<Type const*>` might result in differently ordered types for the same input. It does not matter in a healthy situation. But in case of bugs, we might see different behavior on different systems (or even on the same system), making the picture non-deterministic.

Or is it too theoretical?

Reply:
Yeah, the order will be non-deterministic because the addresses are pretty random. But I personally wouldn't worry about this potentially hiding bugs - it could just as well make them more visible. And I think there's a lot of other non-determinism in the compiler anyway.

I'd either go all the way and guarantee the order or just not depend on the order - and for me returning `set` (or even better, `unordered_set`) makes it much clearer. And if you do want to guarantee order then exposing the comparator might actually be a good idea.

Reply:
I think that after switching to (deterministic) `allSubtypes/flattenedSubtypeTree`, this conversation became obsolete :).